### ISSUE TITLE
Default to -Wall (but not -Werror)

### ISSUE THREAD

**Issue by schmonz** [notqmail/notqmail#189]
1. Get to warning-free builds in our usual dev environments

2. Submit a PR adding `-Wall -Werror` to `conf-cc` (and maybe shrinking our CI build matrix)

3. Test it on [every platform we can get our hands on](https://github.com/notqmail/notqmail/wiki/Platforms)

4. Fix whatever comes up

5. Merge

6. Rejoice!

7. Proceed directly to #48

**Comment by schmonz** [comment]
#83

**Comment by schmonz** [comment]
Must have click tapped something by mistake, this ain’t done yet

**Comment by DerDakon** [comment]
Just to make this clear: this affects only CI builds, no? Otherwise I'll hate you for -Werror as it will break with every new compiler version and any released version.

**Comment by schmonz** [comment]
Hmmm, I had been thinking we’d use it daily and fix whatever comes up, and ship with it and fix whatever users report. So yeah, the default `conf-cc`. Which scenario (or both) are you worried about?

**Comment by DerDakon** [comment]
The point is to have it in the CI so we see everything that happens for us, but not have it in the default that ships to a user as that could easily break if it is built with a different compiler or on a different platform for no good reason. -Wall is fine, but not -Werror.

**Comment by schmonz** [comment]
Heh, it keeps auto-closing this with _each_ linked issue instead of after _all_ of them (or not automatically at all). Guess I’m using this GitHub feature wrong.

I can enable `-Werror` in pkgsrc notqmail, pay attention to bulk builds on all platforms like I already do, and report here (and quickly work around there) whatever platform-specific warnings pop up in each notqmail release. This will give us pretty good platform coverage without setting a precarious default for everyone who builds.

**Comment by schmonz** [comment]
For a data point, gcc 7.5.0 with `-DDEPRECATED_FUNCTIONS_REMOVED=1 -Wall -Werror -Wno-misleading-indentation -Wno-builtin-declaration-mismatch -Wno-pointer-sign -Wno-maybe-uninitialized -Wno-unused-variable` in `conf-cc` completes `make test` on macOS Big Sur.

**Comment by schmonz** [comment]
Dang it, I keep forgetting that linking a PR to an issue makes the issue close when the PR does.

The approach in #211 (incrementally enable specific warnings in `conf-cc`, and maybe also specific warnings-as-errors) seems prudent, and also not mutually exclusive with other approaches (like enabling _all_ warnings and warnings-as-errors by default in our autobuilds).

**Comment by schmonz** [comment]
@skarnet suggests (in another part of the forest) also -Wextra.

**Comment by skarnet** [comment]
Yes, `-Wall -Wextra` is annoying but a good way to make sure you don't miss anything important. (It has saved me a couple times.)

Whatever you do, however, *do not* put `-Werror` back. `-Werror` is only for local and temporary use on a machine you control, when you want to programmatically know if a build has warnings. It has no place in software you publish.

**Comment by schmonz** [comment]
> `-Werror` is only for local and temporary use on a machine you control, when you want to programmatically know if a build has warnings. It has no place in software you publish.

Yep, @DerDakon explained this to me so well that today (which is several years later), I easily explained it back to him.

**Comment by DerDakon** [comment]
What our actual plan is is to put -Werror=something into our CI builds once a given class of problems is fixed. I still feel the urge to shoot everyone on sight who enables -Werror on release builds.
